python/lsst/daf/butler/core/dimensions/sets.py

```diff
@@ -0,0 +1,402 @@
+
+    The constructor and `union` and `intersection` methods for
+    `DimensionUnitSet` take an extra ``optional`` argument that controls
+    whether optional dependencies of all elements are also included in
+    the returned set.  This defaults to `False`.
+    """
+
+    def __repr__(self):
+        return f"DimensionUnitSet({str(self)})"
+
+    def _populate(self, names, optional=False):
+        expanded = set()
+
+        # recursive local function to add dependencies of dependencies
+        def visit(unit):
+            expanded.add(unit.name)
+            for dep in unit.dependencies(optional=optional):
+                visit(dep)
+
+        for unitName in names:
+            visit(self.universe.units[unitName])
+
+        for dimension in self.universe.units:
+            if dimension.name in expanded:
+                self._elements[dimension.name] = dimension
+                expanded.remove(dimension.name)
+        assert not expanded
+
+    def symmetric_difference(self, other):
+        """Symmetric difference (XOR) is disabled for `DimensionUnitSet`.
+
+        A `DimensionUnitSet` must always include all required dependencies of
+        any of its elements, which rules out a standard symmetric difference.
+        Returning a standard symmetric difference along with needed
+        dependencies that should have been removed would be confusing in at
+        least some contexts, and we don't have a clear use case for that
+        anyway.
+        """
+        raise TypeError("DimensionUnitSet cannot safely implement "
+                        "symmetric_difference.")
+
+    def difference(self, other):
+        """Set difference is disabled for `DimensionUnitSet`.
+
+        A `DimensionUnitSet` must always include all required dependencies of
+        any of its elements, which rules out a standard difference. Returning
+        a standard difference along with needed dependencies that should have
+        been removed would be confusing in at least some contexts, and we
+        don't have a clear use case for that anyway.
+        """
+        raise TypeError("DimensionUnitSet cannot safely implement "
+                        "set difference.")
+
+
+class DimensionJoinSet(DimensionSet):
+    r"""A concrete `DimensionSet` for `DimensionJoin` instances.
+
+    `DimensionJoinSet`\s are indirectly topologically sorted according to the
+    `DimensionUnit`\s they depend on: a `DimensionJoin` "A" appears before "B"
+    if the last of its (topologically sorted) dependencies appear before
+    the last of B's dependencies.  Ties are guaranteed to be broken
+    deterministically.
+    """
+
+    def __repr__(self):
+        return f"DimensionJoinSet({str(self)})"
+
+    def _populate(self, names):
+        for dimension in self.universe._joins:
+            if dimension.name in names:
+                self._elements[dimension.name] = dimension
+                names.remove(dimension.name)
```

Comment:
can `names` be a `frozenset`?

Reply:
Apparently not; I'll transform to `set` at the beginning of this function both to ensure it is one and avoid argument-mutating side-effects.